orchagent/mlagorch.cpp

```diff
@@ -0,0 +1,522 @@
+    }
+    else
+    {
+
+        m_mlagIntfs.insert(if_name);
+
+        //If this is the first MLAG interface added, create isolation group
+        if (m_mlagIntfs.size() == 1)
+            addIslIsolationGroup();
+        else
+            updateIslIsolationGroup(if_name, true);
+
+        //Notify observer
+        update.if_name = if_name;
+        update.is_add = true;
+        notify(SUBJECT_TYPE_MLAG_INTF_CHANGE, static_cast<void *>(&update));
+    }
+    return true;
+
+}
+bool MlagOrch::delMlagInterface(string if_name)
+{
+    MlagIfUpdate update;
+
+    //Delete an unknown MLAG interface
+    if (m_mlagIntfs.find(if_name) == m_mlagIntfs.end())
+    {
+        SWSS_LOG_ERROR("MLAG deletes unknown MLAG interface %s", if_name.c_str());
+    }
+    else
+    {
+        m_mlagIntfs.erase(if_name);
+
+        //If this is the last MLAG interface added, delete isolation group
+        if (m_mlagIntfs.size() == 0)
+            deleteIslIsolationGroup();
+        else
+            updateIslIsolationGroup(if_name, false);
+
+        //Notify observers
+        update.if_name = if_name;
+        update.is_add = false;
+        notify(SUBJECT_TYPE_MLAG_INTF_CHANGE, static_cast<void *>(&update));
+    }
+    return true;
+}
+
+bool MlagOrch::isMlagInterface(string if_name)
+{
+    if (m_mlagIntfs.find(if_name) == m_mlagIntfs.end())
+        return false;
+    else
+        return true;
+}
+
+bool MlagOrch::isIslInterface(string if_name)
+{
+    if (m_isl_name == if_name)
+        return true;
+    else
+        return false;
+}
+
+/* Create isolation group based on MLAG configuration before ICCPd
+ * takes control of the isolation group
+ */
+bool MlagOrch::addIslIsolationGroup()
+{
+    isolation_group_status_t status;
+    string isolate_dst_ports;
+
+    //Need peer link and at least one MLAG interface to create an isolation group
+    if (m_isl_name.empty() || (m_mlagIntfs.size() == 0))
+    {
+        SWSS_LOG_NOTICE("MLAG skips adding isolation group: isl name(%s), num MLAG interface %zu",
+            m_isl_name.empty() ? "invalid" : m_isl_name.c_str(), m_mlagIntfs.size());
+        return false;
+    }
+    //During SwSS warm reboot, isolation group can be created first by ICCP
+    //due to IsoGrpOrch processing info from APP_DB first
+    auto isolation_grp = gIsoGrpOrch->getIsolationGroup(MLAG_ISL_ISOLATION_GROUP_NAME);
+    if (isolation_grp)
+    {
+        m_iccp_control_isolation_grp = true;
+
+        //Register with IsoGrpOrch to receive update when ICCP deletes the group
+        if (!isolation_grp->isObserver(this))
+        {
+            isolation_grp->attach(SUBJECT_TYPE_ISOLATION_GROUP_CHANGE, this);
+            ++m_num_isolation_grp_attach;
+            m_attach_isolation_grp = true;
+            SWSS_LOG_NOTICE("MLAG found ICCP-controlled isolation group. Attach to it");
+        }
+        else
+            SWSS_LOG_NOTICE("MLAG is already attached to ICCP-controlled isolation group");
+        return false;
```

Comment:
From what I understood, it's positive scenario to attach to the existing isolation group. So, returning `false` value (same as for the cases when `addIsolationGroup()` fails or missing peer link) can be confusing. If it's because of the name of the method (`addIsolationGroup()`), I'd suggest to rename it something like `refIsolationGroup()` and return `true` in both case - newly created group and reference update for the existing one.